### DISCUSSION TITLE
rw-eap-sim-id-radius issues (eap: No mutually acceptable types found)

### DISCUSSION THREAD

**Discussion by subhajitchatu** [strongswan/strongswan#1587] [General]
Hi 
I am testing rw-eap-sim-id-radius [(https://github.com/strongswan/strongswan/tree/master/testing/tests/ikev2/rw-eap-sim-id-radius]. My setup is 
UE [ strongswan ] < ---- > EPDG [ strongswan ] < ---- > FreeRadius Server. 
I am using the same configuration files as given in the above link.  ( only thing I have changed is the authentication method of epdg with UE : from pubkey to PSK ) 

I am getting below errors:
**RADIUS SERVER:**
 Auth: (5) Login incorrect (eap: No mutually acceptable types found): [228060123456001] (from client moon port 3 cli 172.26.0.2[4500])

**UE side**:  [ue side error.txt](https://github.com/strongswan/strongswan/files/10938878/ue.side.error.txt)

Please suggest. 

The background is that:
I am trying to develop an epdg using strongswan. For that I need to authenticate UE using EAP-SIM, EAP-AKA or EAP-AKA' methods with a third party AAA server.

Also I wanted to know if strongswan has support for third party AAA authentication using eap-aka or eap-aka' (eap-aka-3gpp2 plugin). I am not able to find any place to set a AAA ip address in the configuration block.

Thanks

**Comment by tobiasbrunner** [comment]
As I already wrote in my email (why did you post here again? or if you planned to do that, why send an email?), the log tells you why you get that error:

```
server requested EAP_SIM authentication (id 0x8B)
EAP method not supported, sending EAP_NAK
```

The _eap-sim_ plugin is apparently not loaded (or incorrectly configured).




> Also I wanted to know if strongswan has support for third party AAA authentication using eap-aka or eap-aka' (eap-aka-3gpp2 plugin). I am not able to find any place to set a AAA ip address in the configuration block.

Please clarify what you mean with this. On the client or the server? If you mean implement EAP-AKA via a third-party module, you'd have to provide a custom plugin that implements the `simaka_card_t` interface (also, there is currently no plugin that implements EAP-AKA', only EAP-AKA is implemented by the existing plugins).  On the server, the _eap-radius_ plugin can delegate authentication to a RADIUS server, which you are apparently already using.

**Comment by subhajitchatu** [comment]
Hi Tobias,

Sorry for posting the same question on both the places. I thought my mail
might go to the spam folder. Anyway I was trying to enable aka-sim in
strongswan as you suggested. I am installing strongswan 5.9.2 on ubuntu
22.04 docker.

1. It was compiled and installed successfully. But while running "ipsec
start"
ipsec start
Starting strongSwan 5.9.2 IPsec [starter]...
ipsec_starter[1760]: Starting strongSwan 5.9.2 IPsec [starter]...

***@***.***:/strongswan-5.9.2# ipsec startipsec_starter[1763]: charon
has died -- restart scheduled (5sec)

Then I checked -------
 $ gdb /usr/libexec/ipsec/charon
Program received signal SIGSEGV, Segmentation fault.
0x00007fae9bb7e756 in addrblock_validator_create () from
/usr/lib/ipsec/plugins/libstrongswan-addrblock.so


After that I also tried to configure

--disable-addrblock --disable-ldap
--disable-static --enable-eap-sim --enable-eap-sim-file
--enable-eap-simaka-pseudonym --enable-eap-simaka-reauth
--enable-eap-simaka-sql

But can't able to start ipsec

2. Then I tried with strongswan 5.9.9, still not able to start ipsec,
charon restarted and seems it stuck somewhere,

now error coming in charon saying
(gdb) bt
#0  __futex_abstimed_wait_common (cancel=false, private=0, abstime=0x0,
clockid=0, expected=2,
    futex_word=0x55c0b9f202a0) at ./nptl/futex-internal.c:103
#1  __GI___futex_abstimed_wait64 ***@***.***=0x55c0b9f202a0,
***@***.***=2,
    ***@***.***=0, ***@***.***=0x0,
***@***.***=0) at ./nptl/futex-internal.c:128
#2  0x00007f6125b40395 in __pthread_rwlock_wrlock_full64 (abstime=0x0,
clockid=0, rwlock=0x55c0b9f20298)
    at ./nptl/pthread_rwlock_common.c:829
#3  ___pthread_rwlock_wrlock (rwlock=0x55c0b9f20298) at
./nptl/pthread_rwlock_wrlock.c:26
#4  0x00007f6125dc4f7e in write_lock (this=<optimized out>) at
threading/rwlock.c:142
#5  0x00007f6125d959ad in add_entry (this=0x55c0b9f2f5f0,
list=0x55c0b9f1f750, ***@***.***=0,
    plugin_name=<optimized out>, ***@***.***=0, create=<optimized
out>) at crypto/crypto_factory.c:547
#6  0x00007f6125d95b2f in add_nonce_gen (this=<optimized out>,
plugin_name=<optimized out>, create=<optimized out>)
    at crypto/crypto_factory.c:910
#7  0x00007f6125dafaae in plugin_feature_load (plugin=<optimized out>,
feature=0x7f6125992058, reg=0x7f6125992040)
    at plugins/plugin_feature.c:542
#8  0x00007f6125dae07f in load_feature (level=2, provided=0x55c0b9f4f300,
this=0x55c0b9f1ac10)
    at plugins/plugin_loader.c:803
#9  load_provided ***@***.***=0x55c0b9f1ac10, provided=0x55c0b9f4f300,
***@***.***=1)
    at plugins/plugin_loader.c:878
#10 0x00007f6125dadecb in load_provided (level=1, provided=<optimized out>,
this=0x55c0b9f1ac10)
    at plugins/plugin_loader.c:849
#11 load_registered (registered=<optimized out>, registered=<optimized
out>, level=1, this=0x55c0b9f1ac10)
    at plugins/plugin_loader.c:705
#12 load_dependencies (level=<optimized out>, provided=<optimized out>,
this=<optimized out>)
    at plugins/plugin_loader.c:747

3. Sorry for the confusion, I want to build an open source epdg using
strongswan and authentication using third party AAA (UE is simulated using
ipsec client) , for that I thought that epdg will need ikev2 support and it
will carry EAP things in it's payload, strongswan can connect to radius
client/proxy, it will send EAP messages to it, which internally get
transferred to AAA server, After authentication flows are completed, UE
will send final IKE AUTH req, I have to inform PGW to create session and
give UE ip, here I may need to add hooks inside strongswan/charon to hold
the auth for sometime, get things done from the PGW and then return UE the
ip address along with IKE AUTH response. Is this feasible using current
stronswan implementation? Please guide.

Regards,
Subhajit

**Reply by tobiasbrunner** [reply]
You might be mixing plugins from different builds. If you build from sources, make sure to remove all binary packages and plugins from previous builds.

**Comment by subhajitchatu** [comment]
The issue was coming because the docker where I was running ipsec didn't
have NET_ADMIN configured. Issue was found while using the "ipsec start
--attach-gdb" command.
I am trying to build strongswan from source because there are no plugin
eap-* plugins available by-default.

Now I am getting


*1. Configure step showing eap-* is enabled and will be built. (Makefile
generated)*


* strongSwan will be built with the following plugins*
-----------------------------------------------------
libstrongswan: aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation
constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8
fips-prf gmp curve25519 xcbc cmac hmac kdf gcm drbg
libcharon:     attr kernel-netlink resolve socket-default stroke vici
updown eap-identity eap-sim eap-sim-file eap-aka eap-aka-3gpp2
eap-simaka-sql eap-simaka-pseudonym eap-simaka-reauth eap-md5 eap-gtc
eap-radius eap-tls eap-ttls eap-peap eap-tnc xauth-generic counters
libtnccs:      tnc-tnccs
libtpmtss:

*2. This is a fresh make of the library. Successfully built all the
plugins. For eg. EAP-SIM*
***@***.***:/strongswan-5.9.10# ls
src/libcharon/plugins/eap_sim/.libs/
eap_sim_peer.o             eap_sim_server.o
libstrongswan-eap-sim.lai
eap_sim_plugin.o           libstrongswan-eap-sim.la
libstrongswan-eap-sim.so

*3. Make install.*
*All eap-*.conf are there in below paths with " load = yes " *

/usr/share/strongswan/templates/config/plugins/
/usr/local/strongswan.d/charon
/etc/strongswan.d/charon/


*strongswan.conf showing " load_modular = yes " and including all conf
files.*

charon {
        load_modular = yes
        plugins {
                include strongswan.d/charon/*.conf
        }
}

include strongswan.d/*.conf


libraries are also there


***@***.***:/strongswan-5.9.10# ls
/usr/lib/ipsec/plugins/libstrongswan-eap-**libstrongswan-eap-aka-3gpp2.la
      libstrongswan-eap-sim.a
libstrongswan-eap-aka-3gpp2.so         libstrongswan-eap-sim.la
libstrongswan-eap-aka.la               libstrongswan-eap-sim.so
libstrongswan-eap-aka.so               libstrongswan-eap-simaka-pseudonym.la
libstrongswan-eap-gtc.la               libstrongswan-eap-simaka-pseudonym.so
libstrongswan-eap-gtc.so               libstrongswan-eap-simaka-reauth.la
libstrongswan-eap-identity.la          libstrongswan-eap-simaka-reauth.so
libstrongswan-eap-identity.so          libstrongswan-eap-simaka-sql.la
libstrongswan-eap-md5.la               libstrongswan-eap-simaka-sql.so
libstrongswan-eap-md5.so               libstrongswan-eap-tls.la
libstrongswan-eap-peap.la              libstrongswan-eap-tls.so
libstrongswan-eap-peap.so              libstrongswan-eap-tnc.la
libstrongswan-eap-radius.la            libstrongswan-eap-tnc.so
libstrongswan-eap-radius.so            libstrongswan-eap-ttls.la
libstrongswan-eap-sim-file.la          libstrongswan-eap-ttls.so
libstrongswan-eap-sim-file.so

*4. ipsec start --attach-gdb*

Started successfully but plugins not loaded showing

00[DMN] Starting IKE charon daemon (strongSwan 5.9.10, Linux
4.15.0-204-generic, x86_64)
[tcsetpgrp failed in terminal_inferior: Invalid argument]

*00[LIB] plugin 'eap-aka': failed to load - eap_aka_plugin_create not found
and no plugin file available*00[LIB] plugin 'eap-aka-3gpp2': failed to load
- eap_aka_3gpp2_plugin_create not found and no plugin file available
00[LIB] plugin 'eap-gtc': failed to load - eap_gtc_plugin_create not found
and no plugin file available
00[LIB] plugin 'eap-identity': failed to load - eap_identity_plugin_create
not found and no plugin file available
00[LIB] plugin 'eap-md5': failed to load - eap_md5_plugin_create not found
and no plugin file available
00[LIB] plugin 'eap-peap': failed to load - eap_peap_plugin_create not
found and no plugin file available
00[LIB] plugin 'eap-radius': failed to load - eap_radius_plugin_create not
found and no plugin file available
00[LIB] plugin 'eap-sim': failed to load - eap_sim_plugin_create not found
and no plugin file available
00[LIB] plugin 'eap-sim-file': failed to load - eap_sim_file_plugin_create
not found and no plugin file available
00[LIB] plugin 'eap-simaka-pseudonym': failed to load -
eap_simaka_pseudonym_plugin_create not found and no plugin file available
00[LIB] plugin 'eap-simaka-reauth': failed to load -
eap_simaka_reauth_plugin_create not found and no plugin file available
00[LIB] plugin 'eap-simaka-sql': failed to load -
eap_simaka_sql_plugin_create not found and no plugin file available
00[LIB] plugin 'eap-tls': failed to load - eap_tls_plugin_create not found
and no plugin file available
00[LIB] plugin 'eap-tnc': failed to load - eap_tnc_plugin_create not found
and no plugin file available
00[LIB] plugin 'eap-ttls': failed to load - eap_ttls_plugin_create not
found and no plugin file available
00[LIB] plugin 'openssl': failed to load - openssl_plugin_create not found
and no plugin file available
00[LIB] plugin 'tnc-tnccs': failed to load - tnc_tnccs_plugin_create not
found and no plugin file available


*5. ipsec statusall*

 ***@***.***:/strongswan-5.9.10# ipsec statusall
Status of IKE charon daemon (strongSwan 5.9.10, Linux 4.15.0-204-generic,
x86_64):
  uptime: 5 seconds, since Mar 14 02:57:53 2023
  malloc: sbrk 1077248, mmap 0, used 343984, free 733264
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0,
scheduled: 0
  loaded plugins: charon aes des rc2 sha2 sha1 md5 mgf1 random nonce x509
revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem
pkcs8 fips-prf gmp curve25519 xcbc cmac hmac kdf gcm drbg attr
kernel-netlink resolve socket-default stroke vici updown xauth-generic
counters
Listening IP addresses:
  172.17.0.2
Connections:
Security Associations (0 up, 0 connecting):
  none

Please guide me if I am missing anything to load eap related plugins.

**Reply by tobiasbrunner** [reply]
Apparently, you messed something up. Can't say much else without `Dockerfile` to see what exactly you are doing.

**Comment by subhajitchatu** [comment]
Thank you. The issue has been resolved after installing from scratch.
Ubuntu docker version : 22.04 and strongswan 5.9.9.

Regards,
Subhajit
